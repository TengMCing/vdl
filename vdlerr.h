//
// Created by Patrick Li on 9/11/22.
//

#ifndef VDL_VDLERR_H
#define VDL_VDLERR_H

#include <stdio.h>
#include <stdlib.h>

/*-----------------------------------------------------------------------------
 |  Error code
 ----------------------------------------------------------------------------*/

/// @description Error code.
typedef enum VDL_ERR
{
    VDL_ERR_RESOLVED           = 0,
    VDL_ERR_STACK_LIMIT        = 1,
    VDL_ERR_NULL_POINTER       = 2,
    VDL_ERR_INDEX_OUT_OF_BOUND = 3,
    VDL_ERR_UNKNOWN_TYPE       = 4,
    VDL_ERR_INCOMPATIBLE_TYPE  = 5,
    VDL_ERR_INCORRECT_MODE     = 6,
    VDL_ERR_INCOMPATIBLE_MODE  = 8,
    VDL_ERR_ALLOCATE_FAIL      = 9,
    VDL_ERR_GC_FAIL            = 10,
    VDL_ERR_UNIMPLEMENTED      = 11
} VDL_ERR;

/*-----------------------------------------------------------------------------
 |  Error definition
 ----------------------------------------------------------------------------*/

/// @description Global error
/// @param ERR (VDL_ERR). A predefined error code.
/// @param MSG (const char*). A string to give extra information about the error.
static struct
{
    VDL_ERR ERR;
    const char *MSG;
} VDLINT_GERR = {VDL_ERR_RESOLVED, "\0"};

/// @description Get the global error code.
#define vdl_geterr() ((VDL_ERR){VDLINT_GERR.ERR})
/// @description Get the global error message.
#define vdl_geterrmsg() ((const char *){VDLINT_GERR.MSG})

/*-----------------------------------------------------------------------------
 |  Control error printing
 ----------------------------------------------------------------------------*/

/// @description A global variable to control the error printing.
static int VDL_ERRMSG_ON = 1;

/*-----------------------------------------------------------------------------
 |  Check error
 ----------------------------------------------------------------------------*/

/// @description Check the error code and jump to VDL_EXCEPTION if needed.
#define vdlint_ckerr()                           \
    do {                                         \
        if (VDLINT_GERR.ERR != VDL_ERR_RESOLVED) \
            goto VDL_EXCEPTION;                  \
    } while (0)

/*-----------------------------------------------------------------------------
 |  Except
 ----------------------------------------------------------------------------*/

/// @description Exception handler. Use it as a context manager.
/// @param err (int). The target error code.
#define vdl_catch(err) for (int _except = VDLINT_GERR.ERR != (err); _except < 1; VDLINT_GERR.ERR = VDL_ERR_RESOLVED, _except++)

/// @description Final exception handler. Use it as a context manager.
#define vdl_finally() for (int _except = 0; _except < 1; VDLINT_GERR.ERR = VDL_ERR_RESOLVED, _except++)

/*-----------------------------------------------------------------------------
 |  Exit
 ----------------------------------------------------------------------------*/

/// @description Abort the program.
/// @nobacktrace
_Noreturn static inline void vdl_abort(void)
{
    printf("Program aborted!\n");
    exit(EXIT_FAILURE);
}

#endif//VDL_VDLERR_H
