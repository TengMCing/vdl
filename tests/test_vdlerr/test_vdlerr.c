//
// Created by Patrick Li on 9/11/22.
//

#include "../../vdlerr.h"
#include "../test.h"

int test_ckerr(void);
void test_abort(void);

int main(void)
{
    // echo
    echo("Test vdl_geterr:");
    // expect(True)
    if (vdl_err_getcode() == VDLINT_GERR.CODE)
        test_printf("True");
    else
        test_printf("False");

    // echo
    echo("Test VDL_ERRMSG_ON:");
    // expect(1)
    test_printf("%d", VDL_ERR_MSG_ON);

    // echo
    echo("Test vdlint_ckerr, vdl_catch and vdl_finally:");
    test_ckerr();

    test_abort();
    return 0;
}

int test_ckerr(void)
{
    VDLINT_GERR.CODE = 1;
    vdlint_err_check();
VDL_EXCEPTION:
    // expect(Exception triggered!)
    test_printf("Exception triggered!");
    vdl_catch(2) test_printf("Incorrect!");
    // expect(Handled!)
    vdl_catch(1) test_printf("Handled!");
    // expect(0)
    test_printf("%d", vdl_err_getcode());
    VDLINT_GERR.CODE = 1;
    VDLINT_GERR.MSG  = "This is a mock error";
    // expect(This is a mock error)
    test_printf("%s", vdl_err_getmsg());
    // expect(Finally handled!)
    vdl_finally() test_printf("Finally handled!");
    // expect(0)
    test_printf("%d", vdl_err_getcode());
    return 1;
}

_Noreturn void test_abort(void)
{
    // echo
    echo("Test vdl_abort");
    // exit(1)
    // expect(Program aborted!)
    test_printf_wrapper(vdl_abort());
}
