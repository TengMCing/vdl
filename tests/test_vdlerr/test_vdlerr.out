			Test vdl_geterr:
				Subtest 1: True
			Test VDL_ERRMSG_ON:
				Subtest 1: 1
			Test vdlint_ckerr, vdl_catch and vdl_finally:
				Subtest 1: Exception triggered!
				Subtest 2: Handled!
				Subtest 3: 0
				Subtest 4: This is a mock error
				Subtest 5: Finally handled!
				Subtest 6: 0
			Test vdl_abort
				Subtest 1: Program aborted!
